File: examples/agent_integration.py
#!/usr/bin/env python3
"""Example: Agent capacity check before spawning sub-agent.

This demonstrates how an AI agent can check capacity before spawning
a sub-agent to avoid hitting rate limits.
"""

import subprocess
import json
import sys


def check_capacity(service='claude', threshold=20):
    """Check remaining capacity for service.

    Args:
        service: Service to check ('claude', 'codex', or 'all')
        threshold: Minimum remaining percentage (default: 20%)

    Returns:
        Tuple of (has_capacity: bool, message: str)
    """
    try:
        # Run usage-check with JSON output
        result = subprocess.run(
            ['usage-check', service, '--json'],
            capture_output=True,
            text=True,
            timeout=10
        )

        if result.returncode != 0:
            return False, f"Error: {result.stderr}"

        # Parse JSON output
        data = json.loads(result.stdout)

        # Check each service
        for svc in data['services']:
            if not svc['available']:
                continue

            # Check all metrics (prioritize most restrictive)
            for metric in svc['metrics']:
                if metric['remaining_pct'] < threshold:
                    return False, (
                        f"{svc['name']} low capacity: "
                        f"{metric['name']} only {metric['remaining_pct']}% remaining"
                    )

        return True, "Capacity available"

    except subprocess.TimeoutExpired:
        return False, "Timeout checking capacity"
    except json.JSONDecodeError as e:
        return False, f"Invalid JSON response: {e}"
    except Exception as e:
        return False, f"Error: {e}"


def get_most_restrictive_metric(service='claude'):
    """Get the most restrictive (lowest remaining) metric for a service.

    Args:
        service: Service to check ('claude' or 'codex')

    Returns:
        Tuple of (metric_name: str, remaining_pct: int) or (None, None)
    """
    try:
        result = subprocess.run(
            ['usage-check', service, '--json'],
            capture_output=True,
            text=True,
            timeout=10
        )

        if result.returncode != 0:
            return None, None

        data = json.loads(result.stdout)

        # Find service in results
        for svc in data['services']:
            if svc['name'] == service and svc['available']:
                # Find metric with lowest remaining percentage
                min_metric = min(
                    svc['metrics'],
                    key=lambda m: m['remaining_pct']
                )
                return min_metric['name'], min_metric['remaining_pct']

        return None, None

    except Exception:
        return None, None


def main():
    """Example usage: Check Claude capacity before spawning sub-agent."""

    print("Checking Claude capacity...")

    # Get most restrictive metric
    metric_name, remaining_pct = get_most_restrictive_metric('claude')

    if metric_name is None:
        print("✗ Error checking capacity")
        sys.exit(1)

    print(f"Most restrictive metric: {metric_name} ({remaining_pct}% remaining)")

    # Check if we have capacity (20% threshold)
    has_capacity, msg = check_capacity('claude', threshold=20)

    if has_capacity:
        print(f"✓ {msg}")
        print("Spawning sub-agent...")
        # spawn_subagent()
        sys.exit(0)
    else:
        print(f"✗ {msg}")
        print("Deferring sub-agent spawn")
        sys.exit(1)


if __name__ == '__main__':
    main()

File: examples/agent_integration.ts
/**
 * Complete agentic capacity-management loop.
 *
 * By default this uses an installed `lazyusage` binary. Override with
 * LAZYUSAGE_CMD for repo-local development, for example:
 *
 *   LAZYUSAGE_CMD="bun run lazyusage" bun run examples/agent_integration.ts
 */

interface Metric {
  name: string;
  remaining_pct: number;
  resets: string;
}

interface Service {
  name: string;
  available: boolean;
  source: string | null;
  stale: boolean;
  error: string | null;
  metrics: Metric[];
}

interface Snapshot {
  services: Service[];
}

type Level = "green" | "yellow" | "red";

const CLI = process.env.LAZYUSAGE_CMD ?? "lazyusage";

async function runJson(command: string): Promise<Snapshot> {
  const proc = Bun.spawn(["/bin/sh", "-lc", command], {
    stdout: "pipe",
    stderr: "pipe",
  });

  const [stdout, stderr, exitCode] = await Promise.all([
    new Response(proc.stdout).text(),
    new Response(proc.stderr).text(),
    proc.exited,
  ]);

  if (exitCode !== 0) {
    throw new Error(stderr.trim() || `Command failed: ${command}`);
  }

  return JSON.parse(stdout) as Snapshot;
}

async function getUsage(svc?: string): Promise<Snapshot> {
  const command = svc ? `${CLI} ${svc} --json` : `${CLI} --json`;
  return runJson(command);
}

function tightest(snap: Snapshot, svc?: string): Metric | null {
  const service = snap.services.find((entry) => entry.available && (!svc || entry.name === svc));
  if (!service || service.metrics.length === 0) return null;
  return service.metrics.reduce((a, b) => (a.remaining_pct <= b.remaining_pct ? a : b));
}

function level(remaining: number): Level {
  if (remaining >= 50) return "green";
  if (remaining >= 20) return "yellow";
  return "red";
}

function shouldTrust(service: Service | undefined): boolean {
  return !!service && service.source !== "fallback" && !service.stale;
}

function msUntilReset(resets: string): number {
  const now = new Date();
  const short = resets.match(/^(\d{1,2}):(\d{2})(am|pm)$/i);
  if (short) {
    let hour = parseInt(short[1], 10);
    if (short[3].toLowerCase() === "pm" && hour !== 12) hour += 12;
    if (short[3].toLowerCase() === "am" && hour === 12) hour = 0;
    const target = new Date(now);
    target.setHours(hour, parseInt(short[2], 10), 0, 0);
    if (target <= now) target.setDate(target.getDate() + 1);
    return target.getTime() - now.getTime();
  }
  return 3_600_000;
}

function bestAlternative(snap: Snapshot, avoid: string): string | null {
  const alternatives = snap.services
    .filter((service) => service.available && service.name !== avoid && shouldTrust(service))
    .map((service) => ({
      name: service.name,
      headroom: Math.min(...service.metrics.map((metric) => metric.remaining_pct)),
    }))
    .sort((a, b) => b.headroom - a.headroom);

  return alternatives.length > 0 && alternatives[0].headroom >= 20 ? alternatives[0].name : null;
}

const MAX_RETRIES = 3;

for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
  console.log(`\n--- Attempt ${attempt} ---`);

  const snapshot = await getUsage();
  const service = snapshot.services.find((entry) => entry.name === "claude");
  const metric = tightest(snapshot, "claude");

  if (!metric || !service) {
    console.log("Claude is not available.");
    process.exit(1);
  }

  console.log(`Source: ${service.source ?? "unknown"}${service.stale ? " (stale)" : ""}`);
  if (service.error) console.log(`Fetch warning: ${service.error}`);

  if (!shouldTrust(service)) {
    const alternative = bestAlternative(snapshot, "claude");
    if (alternative) {
      console.log(`Claude data is degraded. Failing over to ${alternative}.`);
      process.exit(0);
    }
  }

  const currentLevel = level(metric.remaining_pct);
  console.log(`Tightest: ${metric.name} ${metric.remaining_pct}% [${currentLevel}]`);

  if (currentLevel === "green") {
    console.log("Full speed. Proceeding.");
    process.exit(0);
  }

  if (currentLevel === "yellow") {
    console.log("Throttling: fewer agents, smaller context, cheaper model.");
    process.exit(0);
  }

  const alternative = bestAlternative(snapshot, "claude");
  if (alternative) {
    console.log(`Failing over to ${alternative}.`);
    process.exit(0);
  }

  if (attempt < MAX_RETRIES) {
    const wait = msUntilReset(metric.resets) + 60_000;
    console.log(`Sleeping ${Math.ceil(wait / 60_000)} min until ${metric.resets}...`);
    await Bun.sleep(wait);
  } else {
    console.log("Max retries reached.");
    process.exit(1);
  }
}
